pkg/operator/cluster/controller.go

```diff
@@ -178,15 +182,27 @@ func (c *ClusterController) onAdd(obj interface{}) {
 func (c *ClusterController) onUpdate(oldObj, newObj interface{}) {
 	oldClust := oldObj.(*rookalpha.Cluster).DeepCopy()
 	newClust := newObj.(*rookalpha.Cluster).DeepCopy()
+
 	if !clusterChanged(oldClust.Spec, newClust.Spec) {
-		logger.Debugf("no updates made in the cluster")
+		logger.Infof("no supported updates made to the cluster")
 		return
 	}
 
 	logger.Infof("updating cluster %s", newClust.Namespace)
-	cluster := newCluster(newClust, c.context)
-	if err := cluster.createInstance(c.rookImage); err != nil {
-		logger.Errorf("failed to update cluster in namespace %s. %+v", newClust.Namespace, err)
+	logger.Infof("old cluster: %+v", oldClust.Spec)
+	logger.Infof("new cluster: %+v", newClust.Spec)
+
+	cluster := newCluster(newClust, oldClust, c.context)
+	err := wait.Poll(updateClusterInterval, updateClusterTimeout, func() (bool, error) {
```

Comment:
here we will retry if there's a failure to update the cluster, however I don't think that's going to be good enough for the ability to resume a failed cluster update orchestration.  For instance, if the operator dies then I believe we'll come back in through the `onAdd` path instead, which doesn't have an old cluster CRD spec to diff against.

I'm thinking what we need to do here is to compare desired cluster CRD state vs. discovered actual cluster state, instead of only relying on having the old cluster CRD spec to compare desired against.  That seems more like a reliable "reconciliation" approach.  Of course, if we have an old cluster CRD we can use that, but augmenting that with a real world discovery to compare with sounds much more robust.

Reply:
@jbw976 the operator is the one that sets the watch for cluster. So when it dies and comes back, it will set a new watch. So you will get all the cluster CRD that were created and execute the `onAdd` for each.

I think you are correct about having to compare with the actual state of the cluster (via ceph cmd).  If the operator is dead and an user modifies the CRD, when the operator comes back, you will only get the update CRD. Which also means that `cluster.createInstance()` has to be idempotent.

Reply:
@jbw976 you are correct, the `onAdd` is called each time the operator starts up and runs through the idempotent orchestration to make sure everything is healthy. The `onUpdate` is only called after that if something updates the crd. So we won't be able to rely on the old crd in the `onUpdate` method

Reply:
2 questions that I don't understand still about CRD watching with the operator kit, @travisn said he'd look into it, but perhaps you also know @kokhang:
1. why is `onAdd` called again when the operator pod restarts and sets a watch on the CRD again?  is it a direct and unconditional callback to `onAdd` or is the watch replaying events? If it is a replay of events, does that also apply to `onUpdate` events? we do not want to be replaying update events from the history as that would impact cluster state.
1. what is the threading model of the add/update/delete callbacks?  are they synchronous? if a callback is running while another event is received, will the kit buffer that event and wait until the current callback has completed before calling the next callback for the buffered event?

Reply:
@jbw976 From my practical experience with watchers, Kubernetes is not replaying events. I believe that when a new watcher object (and informer) is set, it will list all the resources you are watching and call the `onAdd`. So if you start a watcher for pods, you will not get all the update events that these pods have experienced before the watcher object was set.

For your 2nd question, kubernetes creates a queue and add events to it. So if you are processing one add event, the other add event will put into a queue.

Reply:
This file may give you more information: https://github.com/kubernetes/client-go/blob/master/tools/cache/controller.go#L264

Reply:
1) The updates are *not* replayed. Only the latest CRD state is found in the call to `onAdd` after startup
2) The callbacks to `onAdd`, `onUpdate`, and `onDelete` are synchronous. They are queued and executed in order of arrival if there are multiple updates
This behavior was tested in the sample operator on k8s 1.8

Reply:
I have updated this PR to make the operator's orchestration of OSDs both resumeable and idempotent.  If the operator is restarted, it will start it's orchestration efforts by finding the difference between desired state (current CRD version) and actual (discovered) state, then attempting to reconcile the two.  The end result should always be a cluster that is in agreement with desired state even when we don't have an "old" cluster spec to diff against.